pkg/storage/storage.go

```diff
@@ -80,69 +84,121 @@ func (p *storage) GetContainerMetrics(pods ...apitypes.NamespacedName) ([]api.Ti
+	p.storePodMetrics(batch.Pods)
+}
+
+func (p *storage) storeNodeMetrics(nodes []NodeMetricsPoint) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	newNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	prevNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	var nodeCount int
+	for _, nodePoint := range nodes {
 		if _, exists := newNodes[nodePoint.Name]; exists {
 			klog.Errorf("duplicate node %s received", nodePoint.Name)
 			continue
 		}
-		nodeCount += 1
 		newNodes[nodePoint.Name] = nodePoint
+		nodeCount++
+
+		if nodePoint.Timestamp.After(p.nodes[nodePoint.Name].Timestamp) {
+			prevNodes[nodePoint.Name] = p.nodes[nodePoint.Name]
+		} else {
+			prevNodes[nodePoint.Name] = p.prevNodes[nodePoint.Name]
+		}
+		nodeCount++
 	}
+	p.nodes = newNodes
+	p.prevNodes = prevNodes
 
-	newPods := make(map[apitypes.NamespacedName]PodMetricsPoint, len(batch.Pods))
-	for _, podPoint := range batch.Pods {
+	pointsStored.WithLabelValues("node").Set(float64(nodeCount))
+}
+
+func (p *storage) storePodMetrics(pods []PodMetricsPoint) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	newPods := make(map[apitypes.NamespacedName]map[string]ContainerMetricsPoint, len(pods))
+	prevPods := make(map[apitypes.NamespacedName]map[string]ContainerMetricsPoint, len(pods))
+	var containerCount int
+	for _, podPoint := range pods {
 		podIdent := apitypes.NamespacedName{Name: podPoint.Name, Namespace: podPoint.Namespace}
 		if _, exists := newPods[podIdent]; exists {
 			klog.Errorf("duplicate pod %s received", podIdent)
 			continue
 		}
-		containerCount += len(podPoint.Containers)
-		newPods[podIdent] = podPoint
+
+		newContainers := make(map[string]ContainerMetricsPoint, len(podPoint.Containers))
+		prevContainers := make(map[string]ContainerMetricsPoint, len(podPoint.Containers))
+		for _, contPoint := range podPoint.Containers {
+			if _, exists := newContainers[contPoint.Name]; exists {
+				klog.Errorf("duplicate container %s received", contPoint.Name)
+				continue
+			}
+			newContainers[contPoint.Name] = contPoint
+			containerCount++
+
+			if contPoint.Timestamp.After(p.pods[podIdent][contPoint.Name].Timestamp) {
+				prevContainers[contPoint.Name] = p.pods[podIdent][contPoint.Name]
+			} else {
+				prevContainers[contPoint.Name] = p.prevPods[podIdent][contPoint.Name]
+			}
+			containerCount++
+		}
+		newPods[podIdent] = newContainers
+		prevPods[podIdent] = prevContainers
 	}
+	p.pods = newPods
+	p.prevPods = prevPods
 
-	pointsStored.WithLabelValues("node").Set(float64(nodeCount))
 	pointsStored.WithLabelValues("container").Set(float64(containerCount))
-	p.mu.Lock()
-	p.nodes = newNodes
-	p.pods = newPods
-	p.mu.Unlock()
 
 }
+
+func cpuUsageOverTime(metricPoint, prevMetricPoint MetricsPoint) resource.Quantity {
+	if metricPoint.Timestamp.IsZero() || prevMetricPoint.Timestamp.IsZero() {
+		return *resource.NewMilliQuantity(0, resource.DecimalSI)
+	}
+
+	window := math.Max(metricPoint.Timestamp.Sub(prevMetricPoint.Timestamp).Seconds(), 1)
```

Comment:
Why use `math.Max`?

Reply:
It was a legacy check to avoid dividing by zero. I removed it as it not useful anymore